internal/gcsx/random_reader.go

```diff
@@ -339,6 +386,25 @@ func (rr *randomReader) ReadAt(
 		return
 	}
 
+	// Special optimization for random read flow:
+	// Instead of starting a sequential read upto 200MiB, make a gcs request of size GCSFuse receives from kernel.
+	// This stops -
+	// (a) spamming the server because of large 200 MiB request (prefetch).
+	// (b) Head of line blocking issue for gRPC in case multiple reader creates a 200 MiB reader connection and
+	// don't read from that.
+	if !rr.randReadOptimizerState.isDisabled() {
+		if rr.randReadOptimizerState.isFirstRead() {
+			objectData.Size, err = rr.readFromRangeReader(ctx, p, offset, offset+int64(len(p)), util.Random)
+			rr.randReadOptimizerState = randReadOptimizerState(offset + int64(len(p)))
+			return
```

Comment:
Also we should make explicit return here. e.g.` return Obj, nil` or `nil, obj` to make it more readable.